### ISSUE TITLE
Wizard's Tower second floor is boring and tedious

### ISSUE THREAD

**Issue by Ardub23** [copperwater/xNetHack#171]
Granted, I've only played it once, but it was a real hassle that I don't think is befitting of xNetHack. In ordinary mazes, it's easy to know where you haven't explored, and you can `_` travel right to any unexplored paths. Not so in a teleporter maze. So I'd hoped there'd be some generosity in the maze generation algorithm, but that doesn't seem to be the case.

Here are the troubles I faced with the way the maze is generated:

* The maze is just awfully big. Twenty-seven rooms and eighty teleporters is a lot, especially when it takes several steps to reach any particular room. This could be corrected by simplifying the map layout.
* There was only one path to the exit. Out of all eighty teleporters on the floor, only one led to the room with the ladder up. And the room containing that teleporter could only be reached with one other specific teleporter, and so on all the way back to the start. I believe this could be corrected by, after initially generating the maze, redirecting some extra teleporters toward rooms on the correct path (as long as the maze remains [strongly connected](https://en.wikipedia.org/wiki/Strongly_connected_component)).
* The path to the exit was very long. Reaching the ladder up required no less than eleven correct teleporters in a row, and I had to re-tread that path twenty-one times because I had no way of telling it apart from the twenty possible wrong turns along the way. I believe this could be corrected by forcing a room to connect to the exit after a smaller number of steps away from the start.

I ended up creating a map of the floor in a separate program to track teleporters' destinations. Of the 80 teleporters on the floor, the one leading to the end was the 79h one I explored (with the last leading from there to the start). Here is that map; lowercase letters are teleporters, capital letters are their targets. The `^` is a hole.

```
             d#h d#C n#m             
         ##l ###   # ##### q         
       ##K## f###a # n#O## ###P      
    #U h##    m### #   ##e c  # æ    
  g###     c#    # ##d j#    ## ##Z  
  ###  k#h  ###p J j##    d###  #    
 b##v j###o B          M#      ##    
      ##Q##   I#j b d######h c##f #e 
 ###D a###f c##   #   j##f        X# 
 e# #  c#p   #n k#A#d      h###   ## 
    #a     z      #   c##u    # o y# 
 a#    l #####Y f g e  T## ###N #    
  ##E g###      #   #c ##n i    V##w 
  f        #b k#L#o ## ^          #  
    s####R##        F    d#r W##d c  
      n#     #a k i   t h### ####    
         Æ####  # ## ##  #G# x#      
             < a###H n#S             
```

Here's a complete graph of the maze, created with edotor.net. The singular correct path is marked with red arrows; note that each room along the path has only one arrow pointing into it (except G, the second room, which can also be reached by taking a wrong turn later along the path). This means that any room not on that path is even farther from the exit than where you started, and sixty-one out of eighty teleporters lead to those rooms. Also note how the graph is an awful mess, and imagine trying to navigate it blindly.

![image](https://github.com/copperwater/xNetHack/assets/48235540/e09558dc-ebac-483e-8ec7-0019b90676fe)

For comparison, here's a graph of a teleporter maze from another game, the Laverre Gym from _Pokémon X_ and _Y_. This maze has nine rooms and eleven two-way teleporters. If you ignore the ability to back the way you came, no room has more than two teleporters to choose from.

![image](https://github.com/copperwater/xNetHack/assets/48235540/3bb739e4-05de-413e-8ed0-18516aa2aefc)

**Comment by loggersviii** [comment]
In case you aren’t aware, you can see spaces you have and haven’t visited using the #terrain command. I’ve also done this level and it was somewhat helpful

**Comment by Ardub23** [comment]
I'm aware of that, though admittedly I forgot about it while I was doing the maze. Still, it doesn't tell you how to reach rooms that have unexplored teleporters, so I don't believe it's enough to negate the three main issues I mentioned about the maze generation.

**Comment by loggersviii** [comment]
> it doesn't tell you how to reach rooms that have unexplored teleporters

Well, to have gotten to a room, seen a teleporter, and proceed to not step on it implies that you have found a way into that room (obviously), via another teleporter. I think we've already accepted that it's silly to try and do this without taking some form of notes, so as long as you've been keeping track of where teleporters go you'll always be able to get back in a matter of seconds. If you haven't been keeping track of which teleporter you stepped on to get where you are, that's a hell of your own making. 

I think the teleporter maze is fine as long as don't go into it with the delusion that you won't have to take notes of some kind, whether that be pen+paper or in some way digitally. It makes for an awful experience with that delusion, but once I started drawing out the map I actually had quite a bit of fun. IMO you and I had a worse time with the maze because no one had done it before, so we didn't know what to expect. I'd be hesitant on suggesting change from such a small group's experiences.

**Comment by Ardub23** [comment]
My main issue isn't that I had to take notes, though it feels wrong to suddenly need external notes when the game has so many features specifically to counteract that. The issue I want addressed is that, even with my map, I still spent well over an hour on a single floor, where the only gameplay is walking around. For comparison, the previous floor took me fifteen minutes and actually involved problem solving.

When the maze generator admits a maze like the one I got, there's no better way to find the end than an exhaustive search of every possible arbitrary choice, which involves retracing the same paths dozens of times. Making players do that is not in line with xNetHack's stated goals to "eliminate tedious and frustrating parts of the game" and "make uninteresting things more diverse, unique, and varied".

**Comment by k21971** [comment]
I'm in this level now. This is... not fun 🙁 Arbub23, your map either isn't correct, or the to/from locations change each game. I think. My head hurts.

**Comment by Ardub23** [comment]
Yeah, the maze is randomized each game. I included my map to prove that I really did exhaust every single possibility before I found the end.

**Comment by k21971** [comment]
Ooooh ok. I was hoping you'd say the map was right and the maze is the same each game, hah. So this means no making a set map and placing it on the wiki as a spoiler.

**Comment by copperwater** [comment]
Level has been revamped in the 9.0-savebreaking branch. 20 rooms, 32 two-way teleporters, and an algorithm that creates a maze that tries to muddle your path through lots of interconnections rather than finding the One True Solution. Plus stuff in the maze to confuse and slow you down.
